src/lsp.ts

```diff
@@ -0,0 +1,188 @@
+import * as Admzip from "adm-zip";
+import * as which from "which";
+import * as https from "https";
+import * as crypto from "crypto";
+import * as fs from "fs";
+import * as os from "os";
+import * as path from "path";
+import * as vscode from "vscode";
+import { ExtensionContext, WorkspaceConfiguration, workspace } from "vscode";
+import { Executable, LanguageClient, LanguageClientOptions, ServerOptions, TransportKind } from "vscode-languageclient/node";
+
+export abstract class LanguageServerClient {
+  config: WorkspaceConfiguration;
+  ls: LanguageClient | null = null;
+  executableName: string;
+  context: ExtensionContext;
+  private clientOptions: LanguageClientOptions = {
+    documentSelector: ["meson", { "scheme": "file", language: "meson" }]
+  };
+
+  constructor(executableName: string, context: ExtensionContext) {
+    this.executableName = executableName;
+    this.context = context;
+    this.config = workspace.getConfiguration("mesonbuild");
+    const serverModule = this.languageServerPath;
+
+    if (serverModule == null)
+      return;
+    this.startLanguageServer(serverModule);
+  }
+
+  abstract supportsSystem(os: string, arch: string): boolean
+
+  abstract get downloadInfo(): [url: string, hash: string] | null;
+
+  abstract get repoURL(): string;
+
+  restart(): void {
+    this.dispose();
+    const serverModule = this.languageServerPath;
+
+    if (serverModule == null)
+      return;
+    this.startLanguageServer(serverModule);
+  }
+
+  private startLanguageServer(serverModule: string) {
+    const runExe: Executable = {
+      command: serverModule,
+      args: ["--lsp"],
```

Comment:
After you fix the previous comment do the abstract getter for exposing the run and debug exes. The subclasses can access this.languageServerPath and not need to take serverModule as an argument anymore